babel.config.js

```diff
@@ -25,52 +25,213 @@
+  Object.keys(experimentsConfig).forEach(experiment => {
+    const experimentDefine =
+      experimentsConfig[experiment]['defineExperimentConstant'];
+
+    function flagExists(element) {
+      return element['identifierName'] === experimentDefine;
+    }
+
+    // only add default replacement if it already doesn't exist in array
+    if (experimentDefine && !replacements.some(flagExists)) {
+      replacements.push(createReplacement(experimentDefine, false));
+    }
+  });
+
+  // default each backup experiment constant to the customized value
+  for (const [experimentDefine, value] of Object.entries(
+    experimentsConstantBackup
+  )) {
+    function flagExists(element) {
+      return element['identifierName'] === experimentDefine;
+    }
+
+    // only add default replacement if it already doesn't exist in array
+    if (experimentDefine && !replacements.some(flagExists)) {
+      replacements.push(createReplacement(experimentDefine, !!value));
+    }
+  }
+
+  return ['minify-replace', {replacements}];
+}
+
+const eliminateIntermediateBundles = () => [
+  localPlugin('transform-prune-namespace'),
 ];
 
+const getJsonConfigurationPlugin = () =>
+  localPlugin('transform-json-configuration');
+
+/**
+ * Resolves babel plugin set to apply before compiling on singlepass.
+ * @param {!Object<string, boolean>} buildFlags
+ * @return {!Array<string|!Array<string|!Object>>}
+ */
+function plugins({isEsmBuild, isForTesting, isSinglePass, isChecktypes}) {
+  const applied = [...defaultPlugins(isEsmBuild || false)];
+  // TODO(erwinm): This is temporary until we remove the assert/log removals
+  // from the java transformation to the babel transformation.
+  // There is currently a weird interaction where when we do the transform
+  // in babel and leave a bare "string", Closure Compiler does not remove
+  // the dead string expression statements. We cannot just outright remove
+  // the argument of the assert/log calls since we would need to inspect
+  // if the arguments have any method calls (which might have side effects).
+  if (isSinglePass) {
+    applied.push(localPlugin('transform-amp-asserts'));
+  }
+  if (isEsmBuild) {
+    applied.push([
+      'filter-imports',
+      {
+        imports: {
+          ...esmRemovedImports,
+          ...validTransformedRemovableImports,
+        },
+      },
+    ]);
+  }
+  if (isChecktypes) {
+    applied.push(localPlugin('transform-simple-object-destructure'));
+  } else {
+    // This triggers some conformance errors such as `use tryParseJson` during
+    // type check phase so we omit it from the default plugins.
+    applied.push(getJsonConfigurationPlugin());
+  }
+  if (!(isForTesting || isChecktypes)) {
+    applied.push(
+      localPlugin('amp-mode-transformer'),
+      localPlugin('is_dev-constant-transformer')
+    );
+  }
+  return applied;
+}
+
 // eslint-disable-next-line local/no-module-exports
 module.exports = function(api) {
+  const isClosureCompiler =
```

Comment:
These `const` declarations/definitions were moved within the function in part to eliminate the name conflict of `plugins` and in part to make it clear they are used/owned by this entrypoint.

Reply:
SGTM for this PR. (Agree that any further refactoring belongs in a separate PR.)